### PR TITLE
core: typed watch handlers and predicates

### PR DESCRIPTION
**Pull request by jhoblitt**
All existing controller runtime watches are converted to use "typed" handlers and predicates instead of operating on `client.Object`.  The intent is to be bug for bug equivalent with the existing logic while replacing run time type assertions and switch statements with compile time type constraints and type casts. In several cases, functions using assertions were split up such that each function only handles a single Kind at a time. It is hoped that this will improve readability and maintainability while facilitating future refactoring such as migrating some watches to using IndexFields.

Of particular note is that the massive switch statement in `WatchControllerPredicate()`  from `pkg/operator/ceph/controller/predicate.go` has been replaced with generics, reflection, and splitting the obc logic into its own predicate function. There are still many helper functions operating on `client.Object`. These were not updated unless required by the compiler in order to limit the size of this change. The type safety of these funcs should be improved as followup work.

 It is strongly suggested that going forward, handlers and predicates only handle a single Kind (generic or not) and that switches / type assertions are heavily discouraged or forbidden. This PR removed all but a single switch statement in a predicate, which should be addressed in future work.
  

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/cluster/controller.go [hunk 1]
```diff
@@ -190,11 +221,14 @@ func add(opManagerContext context.Context, mgr manager.Manager, r reconcile.Reco
...
 		logger.Info("enabling hotplug orchestration")
-		s := source.Kind[client.Object](
-			mgr.GetCache(),
-			&corev1.ConfigMap{TypeMeta: metav1.TypeMeta{Kind: "ConfigMap", APIVersion: corev1.SchemeGroupVersion.String()}},
-			handler.EnqueueRequestsFromMapFunc(handlerFunc), predicateForHotPlugCMWatcher(mgr.GetClient()))
-		err = c.Watch(s)
+		err = c.Watch(
+			source.Kind(
+				mgr.GetCache(),
+				&corev1.ConfigMap{TypeMeta: metav1.TypeMeta{Kind: "ConfigMap", APIVersion: corev1.SchemeGroupVersion.String()}},
```
Comment (jhoblitt) [inline pkg/operator/ceph/cluster/controller.go:227]: Is there any reason this can't be simplified to `&corev1.ConfigMap{}`?  I doubt there are many pre v1 cm's left in the world...

**Review by travisn — COMMENTED** [review]
Looks good to me, after taking a quick look. This is likely the type of change that if it builds and passes CI, it's likely stable. Anything left from your perspective for test, or ready for merge?

**Comment by jhoblitt** [comment]
> Looks good to me, after taking a quick look. This is likely the type of change that if it builds and passes CI, it's likely stable. Anything left from your perspective for test, or ready for merge?

After discussion with @BlaineEXE , I'm going to simplify the composite literals for the core types. E.g.
```
&corev1.Secret{TypeMeta: metav1.TypeMeta{Kind: "Secret", APIVersion: corev1.SchemeGroupVersion.String()}}
```
to
```
&corev1.Secret{}
```

This should be ready to go in a few minutes.  I do worry that many of the integration tests don't seem to have good coverage for what events trigger reconciliation but I don't think there is much that can be done about that in the short term. I tried hard to make this a fairly mechanical conversion to keep the risk low, even though there are some obvious cleanups that could be done.

**Comment by jhoblitt** [comment]
After spending 20 mins on doing that, I decided to revert back to the the way it was as the usage of `apiutil.GVKForObject()` in `ObjectToCRMapper()` has me concerned that `APIVersion` might important for some controllers (I'm unsure). So I'd like to punt further cleanups to a future PR.

File: pkg/operator/ceph/controller/predicate.go [hunk 2]
```diff
@@ -530,24 +205,23 @@ func WatchPredicateForNonCRDObject(owner runtime.Object, scheme *runtime.Scheme)
 					return true
 				}
 
-				// If the resource is a ConfigMap we don't reconcile
-				_, ok := e.ObjectNew.(*corev1.ConfigMap)
-				if ok {
+				// create a deep copy to evade the type constraint and allow use of .(type)
+				switch newObjCopy := e.ObjectNew.DeepCopyObject().(type) {
```
Comment (jhoblitt) [inline pkg/operator/ceph/controller/predicate.go:209]: It just occurred to me that I can probably cast this to `any()` and remove the pointless deep copy.

File: pkg/operator/ceph/cluster/controller.go [hunk 3]
```diff
@@ -146,41 +161,57 @@ func add(opManagerContext context.Context, mgr manager.Manager, r reconcile.Reco
...
 	}
 
 	// Watch all other resources of the Ceph Cluster
 	for _, t := range objectsToWatch {
-		err = c.Watch(
-			source.Kind[client.Object](
-				mgr.GetCache(), t, handler.EnqueueRequestForOwner(mgr.GetScheme(), mgr.GetRESTMapper(), &cephv1.CephCluster{}),
-				opcontroller.WatchPredicateForNonCRDObject(&cephv1.CephCluster{TypeMeta: ControllerTypeMeta}, mgr.GetScheme())),
-		)
+		err = watchOwnedCoreObject(c, mgr, t)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/controller.go:178]: This function is defined but then only used here. It seems like it would be more straightforward to inline this here the same as the other watches, unless I'm missing something with the looping

File: pkg/operator/ceph/controller/predicate.go (hunk 2 above)
Comment (jhoblitt) [reply pkg/operator/ceph/controller/predicate.go]: Yep. That worked. ;)

File: pkg/operator/ceph/disruption/clusterdisruption/add.go [hunk 4]
```diff
@@ -36,6 +37,69 @@ import (
...
+// Watch for CephBlockPools and enqueue the CephCluster in the namespace
+// Watch for CephFileSystems and enqueue the CephCluster in the namespace
+// Watch for CephObjectStores and enqueue the CephCluster in the namespace
+var objectsToWatch = []client.Object{
+	&cephv1.CephBlockPool{},
+	&cephv1.CephFilesystem{},
+	&cephv1.CephObjectStore{},
+}
+
+func cephClusterPredicate[T client.Object]() predicate.TypedFuncs[T] {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/disruption/clusterdisruption/add.go:49]: I see this called with `cephClusterPredicate[*cephv1.CephCluster]()` below. Is there a reason not to make this a CephCluster type here in the definition like `pdbPredicate` below?

File: pkg/operator/ceph/disruption/clusterdisruption/add.go
Comment (BlaineEXE) [inline pkg/operator/ceph/disruption/clusterdisruption/add.go:1]: @sp98 please double check these changes to the cluster disruption controller and predicates. I think things look good, and CI is passing, but I also don't think these are tested for every corner case in CI. 

You don't have to review the whole PR. If the disruptions changes look good, let me know here or by adding an approving review please. Thank you!

**Review by BlaineEXE — COMMENTED** [review]
I have some questions/clarifications and some minor things that could be improved, but nothing critical blocking merge I don't think. 

I've asked Santosh to double check the disruption budget changes since those are important and also have subtle corner cases.

Let's plan to merge this early tomorrow assuming Santosh says things look good.

File: pkg/operator/ceph/controller/handler.go [hunk 5]
```diff
@@ -19,48 +19,47 @@ package controller
...
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+type listObject interface {
+	runtime.Object
+	metav1.ListInterface
+}
```
[the remark this answers is not in the document]
Comment (jhoblitt) [reply pkg/operator/ceph/controller/handler.go]: Nice find! That looks like it will work.

File: pkg/operator/ceph/controller/predicate.go [hunk 6]
```diff
@@ -51,367 +49,44 @@ const (
...
+			}
+			diff := cmp.Diff(GetSpec(objOld), GetSpec(objNew), resourceQtyComparer)
+			if diff != "" {
+				logger.Infof("CR has changed for %q. diff=%s", objNew.GetName(), diff)
+				return true
+			} else if objectToBeDeleted(objOld, objNew) {
+				logger.Debugf("CR %q is going be deleted", objNew.GetName())
+				return true
+			} else if objOld.GetGeneration() != objNew.GetGeneration() {
+				logger.Debugf("skipping resource %q update with unchanged spec", objNew.GetName())
```
[the remark this answers is not in the document]
Comment (jhoblitt) [reply pkg/operator/ceph/controller/predicate.go:85]: Let me see what can be done.  If nothing else, I looks like it should be possible to get a Kind from the object itself without having to access field with `.GetObjectKind().GroupVersionKind().GroupKind().String()`.

[the remark this answers is not in the document]
Comment (jhoblitt) [reply pkg/operator/ceph/controller/predicate.go]: The all method approach didn't work. It was returning only an empty string.

I tried 3 other variations:

With reflection:
```
2025-04-03 23:31:22.595933 D | ceph-spec: update event from "*v1.CephObjectStore": "tls-test-store"
2025-04-03 23:31:22.595902 D | ceph-spec: skipping resource "*v1.CephObjectStore": "tls-test-store" update event with unchanged spec
```

using apiutil to get the GVK. This required passing in the schema from all callers.
```
2025-04-04 00:50:01.273372 D | ceph-spec: update event for "CephObjectStore": "tls-test-store"
2025-04-04 00:48:58.861999 D | ceph-spec: skipping resource "CephObjectStore": "tls-test-store", update event with unchanged spec
```

Same as above but reporting the namespaced name.
```
2025-04-04 02:58:57.478513 D | ceph-spec: update event for "CephObjectStore": "object-ns/tls-test-store"
2025-04-04 02:58:57.480390 D | ceph-spec: skipping resource "CephObjectStore": "object-ns/tls-test-store", update event with unchanged spec
```

I decided the namespaced name was probably the most useful.

File: pkg/operator/ceph/controller/predicate.go [hunk 7]
```diff
@@ -51,367 +49,44 @@ const (
...
-					logger.Infof("CephCOSIDriver CR has changed for %q. diff=%s", namespacedName, diff)
-					return true
-				} else if objectToBeDeleted(objOld, objNew) {
-					logger.Debugf("CephCOSIDriver CR %q is going be deleted", namespacedName)
-					return true
-				} else if objOld.GetGeneration() != objNew.GetGeneration() {
-					logger.Debugf("skipping CephCOSIDriver resource %q update with unchanged spec", namespacedName)
-				}
-
+			logger.Debug("update event on CephObjectStore CR")
```
[the remark this answers is not in the document]
Comment (jhoblitt) [reply pkg/operator/ceph/controller/predicate.go]: Fixed.

File: pkg/operator/ceph/controller/predicate.go [hunk 8]
```diff
@@ -777,3 +438,20 @@ func DuplicateCephClusters(ctx context.Context, c client.Client, object client.O
...
+	val := reflect.ValueOf(obj)
+
+	// If obj is a pointer, get the element
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+
+	spec := val.FieldByName("Spec")
+	if !spec.IsValid() {
+		logger.Debugf("No Spec field found for object %+v", obj)
```
[the remark this answers is not in the document]
Comment (jhoblitt) [reply pkg/operator/ceph/controller/predicate.go]: I converted it to a warning and added the blurb "this should not happen".

File: pkg/operator/ceph/disruption/clusterdisruption/add.go
Comment (BlaineEXE) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go:1]: I went through this again this morning thoroughly and don't see any code/behavior change. Proceeding to merge this

**Review by BlaineEXE — APPROVED** [review]

File: pkg/operator/ceph/disruption/clusterdisruption/add.go
Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go:1]: I had tested this along with the PR https://github.com/rook/rook/pull/15634.  I didn't see any side effects of the new changes.
